crates/database/src/leader_pg_lock.rs

```diff
@@ -0,0 +1,89 @@
+use {
+    sqlx::postgres::PgPool,
+    std::time::{Duration, Instant},
+};
+
+struct PgLockGuard {
+    conn: sqlx::pool::PoolConnection<sqlx::Postgres>,
+    key: String,
+}
+
+impl PgLockGuard {
+    async fn ping(&mut self) -> bool {
+        sqlx::query("SELECT 1")
+            .execute(&mut *self.conn)
+            .await
+            .is_ok()
+    }
+
+    async fn unlock(mut self) {
+        // TODO what can we do with the result?
+        let _ = sqlx::query("SELECT pg_advisory_unlock($1)")
+            .bind(self.key)
+            .execute(&mut *self.conn)
+            .await
+            .map_err(|err| {
+                tracing::warn!(error = %err, "lock release failed");
+            });
+        // conn returns to pool unlocked
+    }
+}
+
+pub struct LeaderLock {
+    pool: PgPool,
+    key: String,
+    lock_guard: Option<PgLockGuard>,
+    last_try: Instant,
+    try_every: Duration,
+}
+
+impl LeaderLock {
+    pub fn new(pool: PgPool, key: String, try_every: Duration) -> Self {
+        Self {
+            pool,
+            key,
+            lock_guard: None,
+            last_try: Instant::now() - try_every, // allow immediate first try
+            try_every,
+        }
+    }
+
+    // Call every loop; handles acquire/release/liveness. Returns "am I leader rn?"
+    pub async fn tick(&mut self) -> Result<bool, sqlx::Error> {
+        // if we think we're leader, verify the session is alive
+        if let Some(lock) = self.lock_guard.as_mut()
+            && !lock.ping().await
+        {
+            tracing::warn!("leader session died; demoting");
+            self.lock_guard = None; // lock already gone with the dead session
+        }
+
+        // try to become leader if we aren't (rate-limited)
+        if self.lock_guard.is_none() && self.last_try.elapsed() >= self.try_every {
```

Comment:
Still not sure polling is the best option here since `sqlx` has specific support for locks which just wait until the lock is ready.

Reply:
This polls only once in a while on a new block. The wait on the lock would need to happen in a separately spawned task and either set a flag that we got the lock which we would check here or send a message over a channel ("notify") that we would have to have a select!() for here. Why would that be better? What I like about this is that all the code is right here instead of having detached tasks doing stuff which are IMHO harder to find and reason about. 🤷